### PR TITLE
Status annotations

### PR DESCRIPTION
**Pull request by amirmalka**
## Overview

___

## **Type**
enhancement, bug_fix

___

## **Description**
- Introduced handling for dropped events in application profile and network manager.
- Added `k8sObjectCache` to application profile manager for enhanced event processing.
- Implemented new utility methods for watched container status updates.
- Updated error handling to reflect too large application profiles.
- Enhanced tests and mocks to support new changes.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><details><summary>8 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>main.go</strong><dd><code>Refactor Dynamic Watcher Initialization and Start Sequence</code></dd></summary>
<hr>

main.go
<li>Refactored dynamic watcher and k8sObjectCache initialization to avoid <br>duplication.<br> <li> Removed redundant k8sObjectCache creation in the runtime detection <br>section.<br> <li> Moved the dynamic watcher start call to after all adaptors are added.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261">+14/-14</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>applicationprofile_manager_interface.go</strong><dd><code>Interface Update to Support Reporting Dropped Events</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/applicationprofilemanager/applicationprofile_manager_interface.go
- Added `ReportDroppedEvent` method to the interface.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-c8a552f73130367e711a6bd1c8b00dbb9a7b1ab7d7274fc0db6be59f333afb51">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>applicationprofile_manager.go</strong><dd><code>Enhancements to Application Profile Manager for Dropped Events </code><br><code>Handling</code></dd></summary>
<hr>

pkg/applicationprofilemanager/v1/applicationprofile_manager.go
<li>Added <code>k8sObjectCache</code> to <code>ApplicationProfileManager</code> struct.<br> <li> Implemented <code>ReportDroppedEvent</code> to mark events as dropped.<br> <li> Enhanced <code>saveProfile</code> to handle dropped events and update status <br>annotations.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-fc815317651e17975c117749e7661127dbcde82fd9d4d36ebc76cb5b09b3c54e">+66/-7</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>container_watcher.go</strong><dd><code>Handle Dropped Events in Container Watcher</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/containerwatcher/v1/container_watcher.go
<li>Added handling for dropped events in exec, open, and network event <br>callbacks.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-9fa1df18c7f441eb315fcce4daa355a4517ee2d692d1b6a572f93ee5edbd247d">+22/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>network_manager.go</strong><dd><code>Enhancements to Network Manager for Dropped Events Handling</code></dd></summary>
<hr>

pkg/networkmanager/network_manager.go
<li>Added handling for dropped network events.<br> <li> Implemented <code>ReportDroppedEvent</code> to mark network events as dropped.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-cbcd69c2d2fb59b909f8b3cec3ff32348a1ddb4f1d465ece79ead0fa0ba7168b">+82/-63</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>network_manager_interface.go</strong><dd><code>Interface Update to Support Reporting Dropped Events</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/networkmanager/network_manager_interface.go
- Added `ReportDroppedEvent` method to the interface.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-0071d030dc01d58e970e074979804a9df2f4d78f4ea199df561cc6cf9266dcb1">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>applicationprofilecache.go</strong><dd><code>Cache Only Complete Application Profiles</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/objectcache/applicationprofilecache/applicationprofilecache.go
- Added check to only cache complete ("full") application profiles.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-7033af70203ec0c7cabfa58ce543eee2aa32baea58ae8904c87db7c406b51f9b">+13/-6</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>utils.go</strong><dd><code>Utility Enhancements for Watched Container Status Handling</code></dd></summary>
<hr>

pkg/utils/utils.go
<li>Added constants and methods for handling watched container statuses.<br> <li> Updated error constants to reflect too large application profiles.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-81ddbadfb415ccbb9c7af84f11668d1aa5e53c34025bf86d4702f16b4e42f045">+103/-3</a>&nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Tests
</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>applicationprofile_manager_mock.go</strong><dd><code>Mock Implementation of ReportDroppedEvent</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/applicationprofilemanager/applicationprofile_manager_mock.go
- Implemented `ReportDroppedEvent` as a no-operation in the mock.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-ac635659b069359e3de4a50076026861cedfcfe02ac5ebc31176fb5cf49b506e">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>applicationprofile_manager_test.go</strong><dd><code>Update Application Profile Manager Tests for k8sObjectCache</code></dd></summary>
<hr>

pkg/applicationprofilemanager/v1/applicationprofile_manager_test.go
- Updated tests to include `k8sObjectCacheMock`.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-4e4af04b3ed98cb9feaf13f1406a7d71609ab637ea5cb47c4f749cfb240afca1">+3/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>network_manager_mock.go</strong><dd><code>Mock Implementation of ReportDroppedEvent</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/networkmanager/network_manager_mock.go
- Implemented `ReportDroppedEvent` as a no-operation in the mock.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-c1af333278c20e1191d77e9ad0baef0b4fc7ab9a1dab0d83606bb71f4fc1498f">+6/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>network_manager_test.go</strong><dd><code>Update Network Manager Tests for k8sObjectCache</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/networkmanager/network_manager_test.go
- Updated tests to include `k8sObjectCacheMock`.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-e8a17f0b67b47061eaf0c83c3f844cb397d281e1eb7ea9a4cb1b5ee48e9f57e0">+7/-2</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Bug_fix
</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>storage_nocache.go</strong><dd><code>Update Error Handling for Application Profile Patching</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/storage/v1/storage_nocache.go
- Updated error handling to use `TooLargeApplicationProfileError`.

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/212/files#diff-f3093b908b4dcb2e91a2c300f81311dd287f33199f7b7b90fd0de9c686f6a641">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: pkg/utils/utils.go [hunk 1]
```diff
@@ -219,6 +282,23 @@ func (watchedContainer *WatchedContainerData) SetContainerType(wl workloadinterf
 	}
 }
 
+// SetTerminationStatus updates the terminated flag and sets the exit code on the watched container
+func (watchedContainer *WatchedContainerData) GetTerminationExitCode(k8sObjectsCache objectcache.K8sObjectCache, namespace, podName, containerName string) int32 {
+	time.Sleep(3 * time.Second)
+	podStatus := k8sObjectsCache.GetPodStatus(namespace, podName)
+	if podStatus != nil {
```
Comment (amirmalka) [inline pkg/utils/utils.go:289]: TODO: init container should be treated differently (not return 0)

File: pkg/objectcache/applicationprofilecache/applicationprofilecache.go [hunk 2]
```diff
@@ -193,13 +194,19 @@ func (ap *ApplicationProfileCacheImpl) addApplicationProfile(_ context.Context,
...
+	// the cache holds only complete ("full") application profiles.
 	// check if the application profile is completed
-	// TODO: @amir
-
-	// if status was complete and now is not (e.g. mode changed from partial to complete), remove from cache
-	// if ap.slugToAppProfile.Has(apName) {
-	// 	return
-	// }
+	// if status was complete and now is not (e.g. mode changed from complete to partial), remove from cache
+	if completionStatus, ok := appProfile.Annotations[helpersv1.CompletionMetadataKey]; ok {
```
Comment (dwertent) [inline pkg/objectcache/applicationprofilecache/applicationprofilecache.go:200]: @amirmalka  Please add a similar code to the network neighbor object cache

File: pkg/networkmanager/network_manager.go [hunk 3]
```diff
@@ -293,19 +284,27 @@ func (am *NetworkManager) monitorContainer(ctx context.Context, container *conta
...
+				// if exit code is 0 we set the status to completed
+				// TODO: Should we split ContainerHasTerminatedError to indicate if we reached the maxSniffingTime?
+				if watchedContainer.GetTerminationExitCode(am.k8sObjectCache, container.K8s.Namespace, container.K8s.PodName, container.K8s.ContainerName) == 0 {
+					watchedContainer.SetStatus(utils.WatchedContainerStatusCompleted)
+				}
+				am.saveNetworkEvents(ctx, container, watchedContainer)
 				return nil
 			}
+
+			// TODO: @Amir should we handle too large?
```
Comment (amirmalka) [inline pkg/networkmanager/network_manager.go:301]: @dwertent should we support the "too large" status also for NN?

Comment (dwertent) [reply pkg/networkmanager/network_manager.go:301]: We should potentially, but we didn't define it (we did not encounter large NP)

File: pkg/objectcache/k8scache/k8scache.go [hunk 4]
```diff
@@ -94,8 +96,13 @@ func (k *K8sObjectCacheImpl) DeleteHandler(_ context.Context, obj *unstructured.
 		if err != nil {
 			return
 		}
-		k.podSpec.Delete(podKey(pod.GetNamespace(), pod.GetName()))
-		k.podStatus.Delete(podKey(pod.GetNamespace(), pod.GetName()))
+
+		// delete the pod spec and status after 1 minute
```
Comment (dwertent) [inline pkg/objectcache/k8scache/k8scache.go:100]: Why?

Comment (amirmalka) [reply pkg/objectcache/k8scache/k8scache.go:100]: We talked about this... to have the pod status available in case the pod is deleted...

**Review by dwertent — APPROVED** [review]

**Pull request merged**